### ISSUE TITLE
Support OSPF NSSA external default route origination

### ISSUE THREAD

**Issue by arifogel** [batfish/batfish#1644]
Need to support originating a default route into NSSA as external OSPF route.

**Comment by dhalperi** [comment]
@progwriter any update here?

**Comment by progwriter** [comment]
No, need much better support for all the ospf area knobs in VI.

**Comment by jeffkala** [comment]
Running into this currently, is there a workaround or update?

**Comment by arifogel** [comment]
Unfortunately @progwriter is no longer active on this repo.

Can you provide a minimal set of configs that showcase this issue? What route do you expect to see where that isn't showing up in Batfish? What are your reproduction steps?

If needed, you can scrub sensitive information using [netconan](https://github.com/intentionet/netconan), available on pypi.

**Comment by jeffkala** [comment]
At a simplified level:

core-rtr <-------> local-rtr1

core-rtr
```
 router ospf 1
   bfd default
   router-id 10.1.1.1
   area 0.0.0.90 stub no-summary
   default-information originate
 ```
 
 local-rtr1
 ```
router ospf 1
   bfd
   router-id 10.1.1.6
   area 0.0.0.90 stub
 ```
  
  This is simplified if more information is needed I can attempt to provide it.
  
  What I'm noticing is when checking the routes in local-rtr1 it does not have a default route.  Because of this all tracing questions fails with no_route from this router.  Looking at the details of the devices in batfish they accurately knew that core-rtr1 is a ospf NSSA and local-rtr1 is a STUB, just expected that batfish would properly inject that default route into the routes.
  
  Hopefully this is enough details let me know if more are needed.

**Comment by jeffkala** [comment]
@arifogel is this enough detail to proceed?  Happy to help however I can.

**Comment by arifogel** [comment]
Thanks @jeffkala, but I still need a bit more information.

To start, I need to know the vendor language. Is it CIsco IOS? NX-OS? Arista EOS (if EOS, give version)? FRR?

Also, need some more config to be able to test in Batfish. If you already have a snapshot zip you are feeding, I could use that. You may anonymize the files and include only relevant ones (but make sure network wildcards and OSPF area numbers are not garbled).

In particular, what is missing from the config for testing is:
- interface config, especially addressing and any OSPF settings
- area assignment of interface(s)
- any other areas (e.g. does core-rtr also participate in area 0?) necessary for this configuration to work
- any static routes that overlap with networks communicated over OSPF, including default route if applicable

**Comment by jeffkala** [comment]
Sure @arifogel , I'll get as much of the data now and will work on sanitized configs after that.

core-rtr = Arista DCS-7050SX2 running 4.23.6M
local-rtr = Cisco N5K running 7.3(9)N1(1)

local-rtr has no other static routes, it is only part of area 90. The SVIs use `ip ospf network point-to-point` with `ip router ospf 1 area 0.0.0.90`.  Then have the ospf config listed in my initial comment.

core-rtr participates in many areas 90 is one of the stubs, but it is also part of area 0.  The SVI config is also `ip ospf network point-to-point` It then has the ospf config listed above for area 90.  There are a few static routes in this config but no that affect this.  One being a default route pointing outbound.

I'll work on sanitizing the two config and get them added.

**Comment by jeffkala** [comment]
@arifogel here's the two sanitized configs.
[core-rtr1.txt](https://github.com/batfish/batfish/files/7694014/core-rtr1.txt)
[local-rtr1.txt](https://github.com/batfish/batfish/files/7694015/local-rtr1.txt)

**Comment by jeffkala** [comment]
Hi @arifogel are these sanitized configs what you were looking for? Let me know how else I can help.

**Comment by arifogel** [comment]
@jeffkala This should be enough, thanks. Having taken a quick look at the configs, I do see some potential config lines I suspect Batfish might not be handling correctly. I 'll take a deeper look this weekend.

**Comment by arifogel** [comment]
So first problem I see on an EOS device is that you have overlapping area `network` commands. Note the following output from EOS CLI:
```
eos(config-router-ospf)#network 10.0.0.0/8 area 0.0.0.0
eos(config-router-ospf)#network 10.0.1.0/24 area 0.0.0.90
! Network 10.0.1.0/24 overlaps with existing network 10.0.0.0/8 of area 0.0.0.0
```
From a bit more testing with those commands in different orders, it looks like this is just a warning; EOS keeps both statements and uses the most specific prefix to determine area:
```
eos(config-router-ospf)#show ip ospf interface
Ethernet1 is up
  Interface Address 10.0.1.124/24, instance 1, VRF default, Area 0.0.0.90
```

On `core-rtr1[Vlan107]` in the config you gave me, I see:
```
router ospf 1
...
   network 10.8.192.0/21 area 0.0.0.0
...
   network 10.8.192.149/32 area 0.0.0.90
```
Of course these ranges overlap.

When I run the two configs you gave me through batfish and run `bf.q.ospfSessionCompatiblity().answer()` with pybatfish, I get:
```
Interface	VRF	IP	Area	Remote_Interface	Remote_VRF	Remote_IP	Remote_Area	Session_Status
0	core-rtr1[Vlan107]	default	10.8.192.149	0	local-rtr1[Vlan107]	default	10.8.192.150	90	AREA_MISMATCH
1	local-rtr1[Vlan107]	default	10.8.192.150	90	core-rtr1[Vlan107]	default	10.8.192.149	0	AREA_MISMATCH
```

So Batfish is using the first network statement to put `core-rtr1[Vlan107]` in area 0 instead of 90 as would be expected. Thus there is no OSPF session created at all in Batfish.

This is a bug since it doesn't match device behavior.

Looking at the code, Batfish does prefer longer prefixes for setting area, but it does some incorrect stuff with ospf network prefixes (fixed in batfish/batfish#7943)

When I fix the bug, the session comes up:
```
Interface	VRF	IP	Area	Remote_Interface	Remote_VRF	Remote_IP	Remote_Area	Session_Status
0	core-rtr1[Vlan107]	default	10.8.192.149	90	local-rtr1[Vlan107]	default	10.8.192.150	90	ESTABLISHED
1	local-rtr1[Vlan107]	default	10.8.192.150	90	core-rtr1[Vlan107]	default	10.8.192.149	90	ESTABLISHED
```

And the default route is now present:
```
Node | VRF | Network | Next_Hop | Next_Hop_IP | Next_Hop_Interface | Protocol | Metric | Admin_Distance | Tag
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
local-rtr1 | default | 0.0.0.0/0 | interface Vlan107 ip 10.8.192.149 | 10.8.192.149 | Vlan107 | ospfIA | 40 | 110 | None
```
